### DISCUSSION TITLE
cpp/memory/how_to_understand_weak_ptr

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#67] [Announcements]
# cpp/memory/how_to_understand_weak_ptr

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/memory/how_to_understand_weak_ptr.html

**Comment by huangjiawei128** [comment]
std::weak_ptr特点

不影响引用计数：
std::weak_ptr 只是对 std::shared_ptr 管理的资源的弱引用，不增加引用计数。
不影响资源的生命周期。

避免循环引用：
主要用于打破循环引用。循环引用是指两个或多个 shared_ptr 互相引用，导致引用计数永远无法变为零，资源无法释放。

必须提升（lock）使用：
在使用 weak_ptr 管理的资源前，必须调用 lock 方法将其转换为 shared_ptr。
如果资源已经被释放，lock 方法返回一个空的 shared_ptr。

**Reply by huangjiawei128** [reply]
std::weak_ptr 提供了一个名为 expired 的成员函数，用于检查 weak_ptr 管理的资源是否已经被释放。

**Reply by huangjiawei128** [reply]
std::shared_ptr：
管理和调整 强引用计数，它决定了资源（对象）何时被释放。每当一个 std::shared_ptr 被复制或赋值时，强引用计数增加；当 std::shared_ptr 被销毁时，强引用计数减少。

std::weak_ptr：
管理和调整 弱引用计数。std::weak_ptr 不会直接影响资源的生命周期，而是用于观察由 std::shared_ptr 管理的资源。每当一个 std::weak_ptr 被创建、复制或赋值时，弱引用计数增加；当 std::weak_ptr 被销毁时，弱引用计数减少。

资源指针的释放
强引用计数为零时：当所有 std::shared_ptr 对象的强引用计数变为零时，资源（即实际管理的对象）会被释放。这是因为没有 std::shared_ptr 对象再持有该资源，资源不再需要。

控制块的释放
控制块的释放条件：控制块只有在其弱引用计数变为零时才会被释放。控制块的释放过程是当所有 std::shared_ptr 和 std::weak_ptr 对象都离开作用域或被销毁时，控制块的弱引用计数也变为零时，控制块才会被释放。
控制块内的资源指针也会在控制块释放时被释放，因为控制块的析构函数会处理资源指针的释放。

**Comment by eliosteven** [comment]
帅北老师还是太强了，讲的真好
